client/client.go

```diff
@@ -998,9 +1001,50 @@ func (c *Client) updateNodeFromFingerprint(response *cstructs.FingerprintRespons
 		c.config.Node.Resources.Merge(response.Resources)
 	}
 
+	for name, newVal := range response.Drivers {
+		oldVal := c.config.Node.Drivers[name]
+		if oldVal == nil && newVal != nil {
+			c.config.Node.Drivers[name] = newVal
+		} else if newVal != nil && newVal.Detected != oldVal.Detected {
+			c.config.Node.Drivers[name].MergeFingerprintInfo(newVal)
+		}
+	}
 	if nodeHasChanged {
 		c.updateNode()
 	}
+
+	return c.config.Node
+}
+
+// updateNodeFromHealthCheck receives a health check response and updates the
+// node accordingly
+func (c *Client) updateNodeFromHealthCheck(response *cstructs.HealthCheckResponse) *structs.Node {
+	c.configLock.Lock()
+	defer c.configLock.Unlock()
+
+	nodeHasChanged := false
+
+	// update the node with the latest driver health information
+	for name, newVal := range response.Drivers {
+		if newVal == nil {
+			continue
```

Comment:
What about when newVal is nil but oldVal is not here?

Reply:
This is just a safety precaution, every driver should have a corresponding driver info value for a node.